nomt/src/beatree/ops/update/branch_updater.rs

```diff
@@ -337,63 +493,125 @@ impl BranchUpdater {
+        // for two reasons:
+        //
+        // 1. It could often happen that the sequence of KeepChunk are interleaved by Update with only a change
+        // in the node pointers
+        // 2. To avoid keeping all the update information within the BranchOp::KeepChunk because it would require
+        // further allocations
+        let apply_chunk =
+            |builder: &mut BranchNodeBuilder, base_range: Range<usize>, ops_range: Range<usize>| {
+                let n_compressed_left = gauge
+                    .prefix_compressed_items()
+                    .saturating_sub(builder.n_pushed());
+
+                let compressed_end =
+                    std::cmp::min(base_range.start + n_compressed_left, base_range.end);
+
+                builder.push_chunk(
+                    &base.node,
+                    base_range.start,
+                    compressed_end,
+                    ops[ops_range]
+                        .iter()
+                        .filter_map(|op| {
+                            if let BranchOp::Update(pos, pn) = op {
+                                Some((pos - base_range.start, *pn))
+                            } else {
+                                None
+                            }
+                        })
+                        .into_iter(),
+                );
+
+                for pos in compressed_end..base_range.end {
+                    let (key, pn) = self.base.as_ref().unwrap().key_value(pos);
+                    builder.push(key, separator_len(&key), pn.0);
                 }
-                // keep a non-compressed separator
-                (BranchOp::Keep(pos, separator_len), None) => {
-                    let (k, pn) = base.key_value(*pos);
-                    builder.push(k, *separator_len, pn.0);
+            };
+
+        let mut pending_keep_chunk = None;
+        // contains a range within `ops` which define the `pending_keep_chunk`
+        let mut pending_ops_range = None;
+        let mut i = 0;
+        while i < ops.len() {
+            // Check if the chunk could grow.
+            // If yes, then update it and restart the loop on the next operation.
+            // Otherwise, apply the pending chunk and let the same operation be re-evaluated.
+            if pending_keep_chunk.is_some() {
+                match &ops[i] {
+                    // found a insert, apply pending chunk
+                    BranchOp::Insert(_, _) => {
+                        apply_chunk(
+                            &mut builder,
+                            pending_keep_chunk.take().unwrap(),
+                            pending_ops_range.take().unwrap(),
+                        );
+                    }
+                    BranchOp::KeepChunk(from, to) => {
+                        // UNWRAP: pending_keep_chunk has just been checked to be Some
+                        let range = pending_keep_chunk.as_mut().unwrap();
```

Comment:
Is this branch even possible? Under what circumstances would `keep_up_to` give us two `KeepChunk`s in a row?

Reply:
Ah...deletions!